iothub_client/tests/common_dt_e2e/iothubclient_common_dt_e2e.c

```diff
@@ -10,568 +10,728 @@
-        setoption_on_device_or_module(OPTION_X509_PRIVATE_KEY, deviceToUse->primaryAuthentication, "Could not set the device x509 privateKey");
+        LogError("Invalid parameter `twin_data`.");
     }
+    else if (Lock(twin_data->lock) == LOCK_ERROR)
+    {
+        LogError("Lock failed.");
+    }
+    else
+    {
+        LogInfo("_reported_twin_data_reset(): Resetting data.\n");
+        twin_data->received_callback = false;
+        free(twin_data->string_property);
+        twin_data->string_property = _generate_unique_string();
+        twin_data->integer_property = _generate_new_int();
 
-    bool trace = true;
-    setoption_on_device_or_module(OPTION_LOG_TRACE, &trace, "Cannot enable tracing");
+        (void) Unlock(twin_data->lock);
+    }
 }
 
-static void destroy_on_device_or_module()
+static void _reported_twin_data_deinit(REPORTED_TWIN_DATA* twin_data)
 {
-    LogInfo("Beginning to destroy IotHub client handle");
-    if (iothub_deviceclient_handle != NULL)
+    if (twin_data == NULL)
     {
-        IoTHubDeviceClient_Destroy(iothub_deviceclient_handle);
-        iothub_deviceclient_handle = NULL;
+        LogError("Invalid parameter `twin_data`.");
     }
-
-    if (iothub_moduleclient_handle != NULL)
+    else
     {
-        IoTHubModuleClient_Destroy(iothub_moduleclient_handle);
-        iothub_moduleclient_handle = NULL;
+        free(twin_data->string_property);
+        Lock_Deinit(twin_data->lock);
+        free(twin_data);
     }
-    LogInfo("Completed destroy of IotHub client handle");
 }
 
-static void sendreportedstate_on_device_or_module(const char* buffer, DEVICE_REPORTED_DATA *device)
+static char* _malloc_and_fill_reported_payload(const char* string, int num)
 {
-    IOTHUB_CLIENT_RESULT result;
+    size_t length = snprintf(NULL, 0, REPORTED_PAYLOAD_FORMAT, num, string, num, string);
+    char* return_value = (char*) malloc(length + 1);
 
-    size_t bufferLen = strlen(buffer);
-
-    if (iothub_moduleclient_handle != NULL)
+    if (return_value == NULL)
     {
-        result = IoTHubModuleClient_SendReportedState(iothub_moduleclient_handle, (unsigned char *) buffer, bufferLen, reportedStateCallback, device);
+        LogError("malloc failed.");
     }
     else
     {
-        result = IoTHubDeviceClient_SendReportedState(iothub_deviceclient_handle, (unsigned char *) buffer, bufferLen, reportedStateCallback, device);
+        (void)sprintf(return_value, REPORTED_PAYLOAD_FORMAT, num, string, num, string);
     }
 
-    ASSERT_ARE_EQUAL(IOTHUB_CLIENT_RESULT, IOTHUB_CLIENT_OK, result, "IoTHub(Device|Module)Client_SendReportedState failed");
+    return return_value;
 }
 
-static void setdevicetwincallback_on_device_or_module(IOTHUB_CLIENT_DEVICE_TWIN_CALLBACK twinCallback, DEVICE_DESIRED_DATA *device)
+static char* _parse_json_twin_char(const char* twin_payload, const char* full_property_name)
```

Comment:
I recall I split things up because there will need to be differentiation betwen json and cbor eventually.  But I think the same concept could be applied here, having a separate set of functions of similar nature for cbor.